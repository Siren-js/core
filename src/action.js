import { Field } from './field';
import * as coerce from './util/coerce';
import {
    isArray,
    isNonNullObject,
    isNullish,
    isString,
    isUri
} from './util/type-guard';

export * from './field';

export class Action {
    #name;
    #href;
    #class;
    #fields;
    #method;
    #title;
    #type;

    constructor(name, href, options = {}) {
        if (!isString(name)) {
            throw new TypeError('Action.name must be a string');
        }
        if (!isUri(href)) {
            throw new TypeError('Action.href must be a URI');
        }

        const {
            class: actionClass,
            fields,
            method,
            title,
            type,
            ...extensions
        } = options ?? {};

        this.#name = name;
        this.href = href;
        this.class = actionClass;
        this.fields = fields;
        this.method = method;
        this.title = title;
        this.type = type;

        Object.keys(extensions).forEach((key) => {
            this[key] = extensions[key];
        });
    }

    get name() {
        return this.#name;
    }

    get href() {
        return this.#href;
    }

    set href(value) {
        this.#href = coerce.toUriReference(value, this.href);
    }

    get class() {
        return this.#class;
    }

    set class(value) {
        this.#class = coerce.toOptionalStringArray(value, this.class);
    }

    get fields() {
        return this.#fields;
    }

    set fields(value) {
        if (isArray(value)) {
            const [fields] = value
                .filter(Field.isValid)
                .map(Field.of)
                .reduce(
                    ([fields, names], field) => {
                        if (!names.has(field.name)) {
                            fields.push(field);
                            names.add(field.name);
                        }
                        return [fields, names];
                    },
                    [[], new Set()]
                );
            this.#fields = Object.freeze(fields);
        } else if (isNullish(value)) {
            this.#fields = undefined;
        }
    }

    get method() {
        return this.#method;
    }

    set method(value) {
        this.#method = coerce.toOptionalString(value, this.method);
    }

    get title() {
        return this.#title;
    }

    set title(value) {
        this.#title = coerce.toOptionalString(value, this.title);
    }

    get type() {
        return this.#type;
    }

    set type(value) {
        this.#type = coerce.toOptionalMediaTypeString(value, this.type);
    }

    toJSON() {
        const {
            name,
            href,
            class: actionClass,
            fields,
            method,
            title,
            type,
            ...extensions
        } = this;
        return {
            name,
            href,
            class: actionClass,
            fields,
            method,
            title,
            type,
            ...extensions
        };
    }

    static isValid(value) {
        return (
            value instanceof Action ||
            (isNonNullObject(value) &&
                isString(value.name) &&
                isUri(value.href))
        );
    }

    static of(value) {
        if (value instanceof Action) {
            return value;
        }
        const { name, href, ...rest } = value;
        return new Action(name, href, rest);
    }
}
