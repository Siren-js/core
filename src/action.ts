/**
 * Copyright (C) 2020  Dillon Redding
 *
 * See the full notice in the LICENSE file at the top level of the repository.
 */
import * as coerce from './coerce';
import deepFreeze from './deep-freeze';
import Extendable from './extendable';
import { field, Field, isField, isParsableField, ParsedField } from './field';
import { isArray, isMediaTypeString, isRecord, isString, isStringArray, isTypedArray, isUndefined, isUri } from './type-guards';

export * from './field';

export function action(name: string, href: string | URL, options: OptionalActionMembers = {}): ParsedAction {
    const { 'class': actionClass, fields, method, title, type, ...extensions } = options;
    const [indexedFields, parsedFields] = coerceFields(fields);
    return deepFreeze({
        name: coerce.toString(name),
        href: coerce.toUriString(href),
        class: coerce.toOptionalStringArray(actionClass),
        method: coerce.toOptionalString(method),
        title: coerce.toOptionalString(title),
        type: coerce.toMediaTypeString(type),
        fields: parsedFields,
        ...extensions,

        findFieldByName(name: string): ParsedField | undefined {
            return indexedFields.get(name);
        },

        updateField<T>(name: string, value: T): ParsedAction {
            if (!indexedFields.has(name)) {
                return this;
            } else {
                return action(this.name, href, {
                    ...options,
                    fields: parsedFields?.map(field => (field.name === name) ? field.update(value) : field)
                });
            }
        }
    });
}

export type OptionalActionMembers = Extendable & Pick<ParsableAction, 'class' | 'fields' | 'method' | 'title' | 'type'>;

export interface ParsableAction extends Extendable {
    class?: string[] | string;
    fields?: Field[];
    href: string | URL;
    method?: string;
    name: string;
    title?: string;
    type?: string;
}

export interface ParsedAction extends Action {
    /**
     * Searches within the action for a field with the given `name`
     * (case-sensitive).
     * @param name Name of the field to search for
     * @returns The field with the given `name`, or `undefined` if it does not exist.
     */
    findFieldByName(name: string): Field | undefined;

    /**
     * Updates the value of the field with the given `name` to `value`.
     * @param name
     * @param value
     * @returns A new action with the updated field.
     */
    updateField<T>(name: string, value: T): ParsedAction;
}

export interface Action extends Extendable {
    /**
     * Describes the nature of an action based on the current representation.
     */
    readonly class?: readonly string[];

    /**
     * A collection of input controls.
     */
    readonly fields?: readonly Field[];

    /**
     * The URI of the action.
     */
    readonly href: string;

    /**
     * An enumerated attribute mapping to a protocol method. When omitted, the
     * default value is assumed to be `'GET'`.
     */
    readonly method?: string;

    /**
     * A string that identifies the action to be performed.
     */
    readonly name: string;

    /**
     * Descriptive text about the action.
     */
    readonly title?: string;

    /**
     * The encoding type for the request. When omitted and the `fields`
     * attribute is present, the default value is assumed to be
     * `'application/x-www-form-urlencoded'`.
     */
    readonly type?: string;
}

function coerceFields(value: unknown): [Map<string, ParsedField>, ParsedField[] | undefined] {
    const indexedFields = new Map<string, ParsedField>();
    if (isArray(value)) {
        value.forEach(v => {
            if (isParsableField(v)) {
                const { name, ...rest } = v;
                if (!indexedFields.has(name)) {
                    indexedFields.set(name, field(name, rest));
                }
            }
        });
        return [indexedFields, [...indexedFields.values()]];
    } else {
        return [indexedFields, undefined];
    }
}

export function isAction(value: unknown): value is Action {
    return isParsableAction(value) &&
        (isUndefined(value.class) || isStringArray(value.class)) &&
        (isUndefined(value.fields) || (isTypedArray(value.fields, isField) && hasUniqueNames(value.fields))) &&
        (isUndefined(value.method) || isString(value.method)) &&
        (isUndefined(value.title) || isString(value.title)) &&
        (isUndefined(value.type) || isMediaTypeString(value.type));
}

export function isParsableAction(value: unknown): value is ParsableAction {
    return isRecord(value) && isString(value.name) && isUri(value.href);
}

function hasUniqueNames(fields: Field[]): boolean {
    const names = fields.reduce((names, { name }) => names.add(name), new Set<string>());
    return names.size === fields.length;
}
