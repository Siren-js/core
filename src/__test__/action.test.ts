import * as Siren from '../action';

const href = 'http://example.com';

describe('action function', () => {
    it('should coerce name parameter to string', () => {
        const cases = [
            ['foo', 'foo'],
            ['foo bar', 'foo bar'],
            ['', ''],
            [undefined, ''],
            [null, ''],
            [true, 'true'],
            [42, '42'],
            [[true, 42, 'foo'], 'true,42,foo'],
            [{}, '[object Object]']
        ];

        cases.forEach(([value, expected]) => {
            const action = Siren.action(value as string, href);
            expect(action.name).toEqual(expected);
        });
    });

    describe('href parameter', () => {
        it('should coerce to URI string', () => {
            const cases = [
                [href, href],
                ['/orders', '/orders'],
                [new URL(href), `${href}/`],
                [true, '/true'],
                [42, '/42'],
                ['foo', '/foo'],
                [[true, 42, 'foo'], '/true,42,foo'],
                [{}, '/[object%20Object]']
            ];

            cases.forEach(([value, expected]) => {
                const action = Siren.action('create', value as string);
                expect(action.href).toEqual(expected);
            });
        });

        it('should reject invalid URI', () => {
            [null, undefined, 'http://\uFFFF.com'].forEach(value => {
                expect(() => Siren.action('create', value as string)).toThrow(TypeError);
            });
        });
    });

    describe('optional parameter', () => {
        it('should set optional members to undefined when empty', () => {
            const name = 'create';

            const action = Siren.action(name, href);

            expect(action.name).toEqual(name);
            expect(action.href).toEqual(href);
            expect(action.class).toBeUndefined();
            expect(action.fields).toBeUndefined();
            expect(action.method).toBeUndefined();
            expect(action.title).toBeUndefined();
            expect(action.type).toBeUndefined();
        });

        it('should coerce class option to string array or undefined', () => {
            const cases = [
                [['order'], ['order']],
                [['customer', 'info'], ['customer', 'info']],
                [[], []],
                [undefined, undefined],
                [null, undefined],
                [true, ['true']],
                [42, ['42']],
                ['foo', ['foo']],
                [[true, 42, 'foo'], ['true', '42', 'foo']],
                [{}, ['[object Object]']]
            ];

            cases.forEach(([value, expected]) => {
                const action = Siren.action('create', href, { class: value as string[] });
                expect(action.class).toEqual(expected);
            });
        });

        it('should coerce method option to string or undefined', () => {
            const cases = [
                ['', ''],
                ['foo', 'foo'],
                ['foo bar', 'foo bar'],
                [undefined, undefined],
                [null, undefined],
                [true, 'true'],
                [42, '42'],
                [[true, 42, 'foo'], 'true,42,foo'],
                [{}, '[object Object]']
            ];

            cases.forEach(([value, expected]) => {
                const action = Siren.action('create', href, { method: value as string });
                expect(action.method).toEqual(expected);
            });
        });

        it('should coerce title option to string or undefined', () => {
            const cases = [
                ['', ''],
                ['foo', 'foo'],
                ['foo bar', 'foo bar'],
                [undefined, undefined],
                [null, undefined],
                [true, 'true'],
                [42, '42'],
                [[true, 42, 'foo'], 'true,42,foo'],
                [{}, '[object Object]']
            ];

            cases.forEach(([value, expected]) => {
                const action = Siren.action('create', href, { title: value as string });
                expect(action.title).toEqual(expected);
            });
        });

        describe('fields option', () => {
            it('should accept array of parsable fields', () => {
                const action = Siren.action('create', href, {
                    fields: [
                        { name: 'foo' },
                        { name: 'bar', type: 'number' },
                        { name: 'baz', class: ['integer'] },
                        { name: 'qux', class: 'integer' },
                        Siren.field('quz')
                    ] as Siren.Field[]
                });

                expect(action.fields).toHaveLength(5);
                expect(action.fields?.[0]).toHaveProperty('name', 'foo');
                expect(action.fields?.[1]).toHaveProperty('name', 'bar');
                expect(action.fields?.[1]).toHaveProperty('type', 'number');
                expect(action.fields?.[2]).toHaveProperty('name', 'baz');
                expect(action.fields?.[2]).toHaveProperty('class', ['integer']);
                expect(action.fields?.[3]).toHaveProperty('name', 'qux');
                expect(action.fields?.[3]).toHaveProperty('class', ['integer']);
                expect(action.fields?.[4]).toHaveProperty('name', 'quz');
            });

            it('should coerce non-array to undefined', () => {
                [null, undefined, true, 42, 'foo', {}].forEach(value => {
                    const action = Siren.action('create', href, { fields: value as Siren.Field[] });
                    expect(action.fields).toBeUndefined();
                });
            });

            it('should ignore invalid items', () => {
                const action = Siren.action('create', href, {
                    fields: [
                        null,
                        undefined,
                        true,
                        42,
                        'foo',
                        { name: 'foo' }
                    ] as Siren.Field[]
                });

                expect(action.fields).toHaveLength(1);
            });

            it('should remove duplicates', () => {
                const action = Siren.action('create', href, {
                    fields: [
                        { name: 'foo', type: 'number' },
                        { name: 'foo', type: 'date' }
                    ]
                });

                expect(action.fields).toHaveLength(1);
                expect(action.fields?.[0]).toHaveProperty('name', 'foo');
                expect(action.fields?.[0]).toHaveProperty('type', 'number');
            });

            it('should be empty if no valid items', () => {
                const action = Siren.action('create', href, {
                    fields: [
                        null,
                        undefined,
                        true,
                        42,
                        'foo'
                    ] as any // eslint-disable-line @typescript-eslint/no-explicit-any
                });

                expect(action.fields).toHaveLength(0);
            });
        });

        describe('type option', () => {
            it('should accept any valid media type string', () => {
                ['application/json', 'text/html', 'image/png'].forEach(type => {
                    const action = Siren.action('create', href, { type });
                    expect(action.type).toEqual(type);
                });
            });

            it('should coerce to invalid media type string to undefined', () => {
                [undefined, null, true, 42, '', 'foo', [true, 42, 'foo'], {}].forEach(value => {
                    const action = Siren.action('create', href, { type: value as string });
                    expect(action.type).toBeUndefined();
                });
            });
        });

        it('should accept extensions', () => {
            const encoding = 'utf-16';

            const link = Siren.action('create', href, { encoding });

            expect(link.encoding).toEqual(encoding);
        });
    });
});

describe('ParsedAction.findFieldByName', () => {
    it('should return field when present', () => {
        const name = 'foo';
        const field = Siren.field(name);
        const action = Siren.action('create', href, { fields: [field] });

        const result = action.findFieldByName(name);

        expect(result?.name).toEqual(name);
    });

    it('should return undefined when absent', () => {
        const action = Siren.action('create', href);

        const result = action.findFieldByName('foo');

        expect(result).toBeUndefined();
    });
});

describe('ParsedAction.updateField', () => {
    it('should return new action with updated field', () => {
        const name = 'foo';
        const action = Siren.action('create', href, {
            fields: [{ name, value: 'bar' }]
        });
        const updatedValue = 'baz';

        const result = action.updateField(name, updatedValue);

        expect(result.name).toEqual(action.name);
        expect(result.href).toEqual(action.href);
        expect(result.fields?.length).toEqual(action.fields?.length);
        expect(result.fields?.[0].name).toEqual(action.fields?.[0].name);
        expect(result.fields?.[0].value).toEqual(updatedValue);
    });

    it('should return same action when field is absent', () => {
        const action = Siren.action('create', href, {
            fields: [
                { name: 'foo' }
            ]
        });

        const result = action.updateField('bar', 'baz');

        expect(result).toBe(action);
    });

    it('should return same action when fields are absent', () => {
        const action = Siren.action('create', href);

        const result = action.updateField('bar', 'baz');

        expect(result).toBe(action);
    });
});

describe('action type guard', () => {
    it('should return true for valid actions', () => {
        const values: Siren.Action[] = [
            // minimal
            { name: 'create', href },
            // kitchen sink
            {
                name: 'create',
                href,
                method: 'POST',
                class: ['order'],
                title: 'Create Order',
                type: 'multipart/form-data',
                fields: [
                    { name: 'orderNumber', type: 'hidden', value: '42' },
                    { name: 'productCode' },
                    { name: 'quantity', type: 'number' }
                ]
            },
            // with extension
            { name: 'create', href, encoding: 'utf-16' }
        ];

        values.forEach(value => {
            expect(Siren.isAction(value)).toEqual(true);
        });
    });

    it('should return false for invalid actions', () => {
        const values = [
            // non-records
            undefined, null, true, 42, 'foo', [],
            // empty record
            {},
            // invalid name
            { name: 42, href },
            // invalid href
            { name: 'create', href: 'http://\uFFFF.com' },
            // invalid class
            { name: 'create', href, class: 'order' },
            // invalid fields
            { name: 'create', href, fields: 'orderNumber' },
            { name: 'create', href, fields: [null] },
            { name: 'create', href, fields: [42] },
            { name: 'create', href, fields: ['orderNumber'] },
            // duplicate field names
            {
                name: 'create',
                href,
                fields: [
                    { name: 'orderNumber', type: 'hidden', value: '42' },
                    { name: 'orderNumber', type: 'number' }
                ]
            },
            // invalid method
            { name: 'create', href, method: 42 },
            // invalid title
            { name: 'create', href, title: 42 },
            // invalid type
            { name: 'create', href, type: 'order' }
        ];

        values.forEach(value => {
            expect(Siren.isAction(value)).toEqual(false);
        });
    });
});
