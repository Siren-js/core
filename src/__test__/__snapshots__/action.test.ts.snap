// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action serialization 1`] = `
"{
  \\"name\\": \\"add-item\\",
  \\"href\\": \\"http://example.com\\",
  \\"class\\": [
    \\"item\\"
  ],
  \\"method\\": \\"POST\\",
  \\"title\\": \\"Add Item\\",
  \\"type\\": \\"text/plain\\",
  \\"fields\\": [
    {
      \\"name\\": \\"orderNumber\\",
      \\"type\\": \\"hidden\\",
      \\"value\\": \\"42\\"
    },
    {
      \\"name\\": \\"productCode\\",
      \\"required\\": true
    },
    {
      \\"name\\": \\"quantity\\",
      \\"class\\": [
        \\"integer\\"
      ],
      \\"type\\": \\"number\\",
      \\"min\\": 1
    }
  ],
  \\"encoding\\": \\"utf-8\\"
}"
`;
